### ISSUE TITLE
Learning Rate and Warmup Steps

### ISSUE THREAD

**Issue by maggieezzat** [google-research/bert#649]
I'm training from the scratch using data of size 720 million words, and sequence length 128, batch size 896, that means 1 epoch is approximately 6000 steps, I have a problem with the learning rate and warmup steps, I tried using 90,000 steps and 10,000 warmup steps with learning rate 1e-4, however, the loss keeps oscillating between 2.5 and 2 and doesn't decrease beyond that. Also the masked lm accuraccy doesn't exceed 60%. I also tried warmup steps 20,000 and I got slightly better loss, which keeps oscillating between 1.3 and 2 and masked lm accuracy 63%. I don't know if the problem is with the learning rate or the warmup steps. Also what was the masked lm accuracy that you got when training the bert base ?

**Comment by maggieezzat** [comment]
@jacobdevlin-google if you can please help with that

**Comment by kwanhong66** [comment]
@maggieezzat Have you had any better result? 
The data size is similar with yours and sequence length is 256 and batch size is 256.
I started training with 10,000 warmup steps and 1e-4 learning rate. The loss decreased from 2.5 to 2 and is still same. Also the masked lm accuracy is about 60%.
I will try training with more warmup steps and lower learning rate but I am not sure if it will make notable changes.
Is there any advice can you give?

**Comment by maggieezzat** [comment]
The masked lm loss will decrease if you continue training, but the best I got was 1.4, however you should check which steps give you better results on your finetuning task, at 90K steps I got masked lm loss 2 and accuracy 63% but got better finetuning results than at 165K steps with masked lm accuracy 70% and loss 1.4

**Comment by biuleung** [comment]
@maggieezzat Have you tried to fine tune your pre-trained model? I got some error messages about the 'key output_bias' and 'output_weights' not found in checkpoint. I then add one more code 'tf.variable_scope("cls/seq_relationship"):' before the declaration of output_weights and output_bias respectively in run_classifier.py, so the fine tune process can be able to find it. I am not sure it is a proper way to do so.
How do you handle it?

**Comment by mahdirezaey** [comment]
hello @maggieezzat  , have you down BERT(MLM , from scratch) on Arabic corpus? 

why warmup is required for  BERT(MLM , from scratch) ?  the idea of cosine annealing is some thing in contradiction with that , i think 
 
i tried even Lr of 1e-2 and Lr did not over shoot  , why is that ? 
have you tried way of  several epoch with 1e-2 then several epoch with 1e-3 , then  1e-4  then  1e-5 ?

**Comment by mahdirezaey** [comment]
in first try , you have 90,000 steps altogether ,  and 10,000 of them were warmup steps ?

**Comment by mahdirezaey** [comment]
> 
> 
> I'm training from the scratch using data of size 720 million words, and sequence length 128, batch size 896, that means 1 epoch is approximately 6000 steps, I have a problem with the learning rate and warmup steps, I tried using 90,000 steps and 10,000 warmup steps with learning rate 1e-4, however, the loss keeps oscillating between 2.5 and 2 and doesn't decrease beyond that. Also the masked lm accuraccy doesn't exceed 60%. I also tried warmup steps 20,000 and I got slightly better loss, which keeps oscillating between 1.3 and 2 and masked lm accuracy 63%. I don't know if the problem is with the learning rate or the warmup steps. Also what was the masked lm accuracy that you got when training the bert base ?

hello @maggieezzat , have you down BERT(MLM , from scratch ,using data of size 720 million words) on Arabic corpus or ?
